### PR TITLE
picoCDN Node VM Helm Chart for OpenNESS

### PR DESCRIPTION
**Pull request by piconets-engineering**
picoCDN Node VM Helm Chart for Intel OpenNESS with supporting files.

### CODE REVIEW

File: applications/picocdn-node-vm/README.md [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+## Pre Requisites - Installing OpenNESS
+
+|**Configuration**                   |                       |
+|  ----------------------------------|-----------------------|
+| OpenNESS Version                   | 20.06                 |
+| Flavor Used 			     | kubevirt VM           |
+| Distribution	                     | IDO       	     |
+
+This document assumes that OpenNESS was installed through [OpenNESS playbooks](https://github.com/open-ness/specs/blob/master/doc/getting-started/network-edge/controller-edge-node-setup.md). Certain configurations had been changed to support VM deployment on OpenNESS for network edge following [openness-network-edge-vm-support.md](https://github.com/open-ness/specs/blob/master/doc/applications-onboard/openness-network-edge-vm-support.md).
```
Comment (cjnolan) [inline applications/picocdn-node-vm/README.md:40]: The supported OpenNESS version is 20.06, could you update the links to the OpenNESS documentation to point to the 20.06 version please?

File: applications/picocdn-node-vm/templates/cdn-service.yaml [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+#  ports:
+#   -  port: {{ default "80" .Values.cdnService.port }}
+#{{- if .Values.cdnService.type }}
+#{{- if eq .Values.cdnService.type "NodePort" }}
+#      nodePort: {{ default "30007" .Values.cdnService.nodePort }}
+#{{- end }}
+#{{- else }}
+#      nodePort: {{ default "30007" .Values.cdnService.nodePort }}
+#{{- end }}
```
Comment (cjnolan) [inline applications/picocdn-node-vm/templates/cdn-service.yaml:23]: Are these commented lines still needed?

File: applications/picocdn-node-vm/templates/cdn-service.yaml [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+apiVersion: v1
+kind: Service
+metadata:
+  name: {{ template "picocdn-node-vm.fullname" . }}-service
+spec:
+  type: {{ default "NodePort" .Values.cdnService.type }}
+  selector:
+    kubevirt.io/domain: {{ template "picocdn-node-vm.fullname" . }}
+  externalIPs:
+    192.168.1.100
```
Comment (archie951) [inline applications/picocdn-node-vm/templates/cdn-service.yaml:13]: '-' sign needs to be added before IP. 
for eg: https://kubernetes.io/docs/concepts/services-networking/service/#external-ips

File: applications/picocdn-node-vm/README.md [hunk 4]
```diff
@@ -0,0 +1,87 @@
+```text
+SPDX-License-Identifier: Apache-2.0
+Copyright (c) 2021 NetPico LAbs Pvt Ltd.(picoNETS)
+```
+
+
+# **picoCDN Platform Overview**
+picoCDN is a Deep Edge focused Caching and Content Delivery Network (CDN) platform. Our Deep Edge CDN platform performs better than traditional CDNs by virtue of it unique and flexible architecture. We achieve this by bringing your content closer to your users, embedding deep into Telco and ISP networks.  With our Ultra-Low Latency CDN platform, we can deliver a buffer-free experience for Full HD, 4K, AR, VR, and other bandwidth-demanding content. Every edge location of our distribution network is geared to deliver content to users who are just a couple of hops away. We also provide a pre-caching API to push your content to the edge nodes before customers requests reach the node. This version of the platform is designed to help you deploy Edge POPs in various local ISPs, malls, hotels, airports, cellsites or other 5G Multi-Access Edge Computing (MEC) locations running Intel OpenNESS.
```
Comment (abjyoti) [inline applications/picocdn-node-vm/README.md:8]: Consider re-writing:  "Avoid using We/Me/You"
picoCDN is a Deep Edge focused Caching and Content Delivery Network (CDN) platform. This Deep Edge CDN platform performs better than other traditional CDNs by virtue of it unique and flexible architecture. This is achieved by bringing the user content closer to user location, embedding deep into Telco and ISP networks.  With this Ultra-Low Latency CDN platform, can deliver a buffer-free experience for Full HD, 4K, AR, VR, and other bandwidth-demanding content. Every edge location of our distribution network is geared to deliver content to users who are just a couple of hops away. This also provide a pre-caching API to push user content to the edge nodes before customers requests reach the node. This version of the platform is designed to help OEM/ODMs deploy Edge POPs in various local ISPs, malls, hotels, airports, cell sites or other 5G Multi-Access Edge Computing (MEC) locations running Intel OpenNESS.

File: applications/picocdn-node-vm/README.md (hunk 1 above)
Comment (abjyoti) [reply applications/picocdn-node-vm/README.md:40]: Can you ensure this is working with latest OpenNESS release 21.03. The OpenNESS 20.06 release is old and may not be supported.

File: applications/picocdn-node-vm/templates/_helpers.tpl [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+{{/*
+Expand the name of the chart.
+*/}}
+{{- define "picocdn-nocd-vm.name" -}}
+{{- default .Chart.Name .Values.nameOverride | trunc 63 | trimSuffix "-" -}}
+{{- end -}}
+
+{{/*
+Create a default fully qualified app name.
+We truncate at 63 chars because some Kubernetes name fields are limited to this (by the DNS naming spec).
```
Comment (abjyoti) [inline applications/picocdn-node-vm/templates/_helpers.tpl:20]: Rephrase: (THESE ARE COSMETIC CHNAGES - SEE IF POSSIBLE TO RESOLVE THESE COMMENTS)
The app name is truncated to 63 characters as some of the Kuberenetes name fields are limited ti this (by the DNS naming spec).

File: applications/picocdn-node-vm/templates/cdn-service.yaml (hunk 2 above)
Comment (abjyoti) [reply applications/picocdn-node-vm/templates/cdn-service.yaml:23]: Try to cleanup unused and commented lines. But if they are required to be modified for different environments and kept for reference/example. it is ok, but put a comment for the same.

File: applications/picocdn-node-vm/README.md (hunk 1 above)
Comment (abjyoti) [reply applications/picocdn-node-vm/README.md:40]: Please update the Readme.md with the troubleshooting steps to use the 20.06 OpenNESS, if they have to use 20.06.

Comment (rajeevrk) [reply applications/picocdn-node-vm/README.md:40]: We are currently in the process of getting access to a test deployment of 21.03.x at the network builders lab, where we will be validating functionality, especially with the updated Calico CNI. We intend to update and push the modified charts once that testing is complete.

Comment (piconets-engineering) [reply applications/picocdn-node-vm/README.md:40]: Resolving via the following edits :
1. Updated the links to specifically reference the 20.06 tagged versions.

Comment (piconets-engineering) [reply applications/picocdn-node-vm/README.md]: Resolving this Suggestion via the following edits 
1. Updated the relevant links to point to the 20.06 tagged versions of the docs on the openness github repo.
2. Additional clarification added to specifically mention that version 20.06 is the required version.

File: applications/picocdn-node-vm/templates/cdn-service.yaml (hunk 2 above)
Comment (piconets-engineering) [reply applications/picocdn-node-vm/templates/cdn-service.yaml]: These lines were a leftover block from the OpenNESS CDN application template, which we used as an initial baseline for our application. We have removed it, to have a cleaner and more compact file.

File: applications/picocdn-node-vm/templates/cdn-service.yaml (hunk 3 above)
Comment (piconets-engineering) [reply applications/picocdn-node-vm/templates/cdn-service.yaml]: Fixed this yaml formatting error, also added the comment identifying that this is an example IP that needs to be updated prior to helm chart deployment. We already had this item in our Installation section of the Readme.md, so this provides the glue instructions in the file.

File: applications/picocdn-node-vm/README.md (hunk 4 above)
Comment (piconets-engineering) [reply applications/picocdn-node-vm/README.md]: Re-written the description block in a neutral perspective, should now avoid the first/second person grammatical constructs.

File: applications/picocdn-node-vm/templates/_helpers.tpl (hunk 5 above)
Comment (piconets-engineering) [reply applications/picocdn-node-vm/templates/_helpers.tpl]: Re-written the comment in a neutral perspective, should now avoid the first/second person grammatical constructs.

File: applications/picocdn-node-vm/README.md [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+## **Where to Purchase**
+
+The picoCDN custom image can be obtained from the picoNETS support portal. For more information visit [www.piconets.com](https://www.piconets.com) or contact us via email on info@piconets.com
+
+
+## Pre Requisites - Installing OpenNESS
+
+|**Configuration**                   |                       |
+|  ----------------------------------|-----------------------|
+| OpenNESS Version                   | 20.06                 |
```
Comment (abjyoti) [inline applications/picocdn-node-vm/README.md:36]: Can you test with latest OpenNESS release like 21.03 and update the release notes accordingly? 
OpenNESS 20.06 is old and will not be supported.

**Review by abjyoti — COMMENTED** [review]
Please resolve the comments

File: applications/picocdn-node-vm/README.md (hunk 6 above)
Comment (rajeevrk) [reply applications/picocdn-node-vm/README.md:36]: As mentioned in the earlier comment, We are currently in the process of getting access to a test deployment of 21.03.x at the network builders lab, where we will be validating functionality, especially with the updated Calico CNI. We intend to update and push the modified charts once that testing is complete.

_Originally posted by @rajeevrk in https://github.com/open-ness/edgeapps/pull/58#discussion_r627397888_

Also, The current customer deployment of this platform is on 20.06, and we have not yet heard from them or Intel on any specific upgrade plans.

**Comment by ranjmish** [comment]
> Please resolve the comments

Hi Anand,
As discussed, it is at Trilogy infrastructure where multiple parties are involve in rural cloud initiative. Piconet team along with Trilogy and Intel have put lot of effort to successful test Piconet app. 

Upgrading openNESS to 21.03 would take time and we are on top of it.  In future, we will test all apps and accordingly update it.
If there is no further review comments, please merge it.

Thank you.

--Ranjan

**Review by abjyoti — DISMISSED** [review]
Approved based on the request

**Review by archie951 — APPROVED** [review]

File: applications/picocdn-node-vm/README.md [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+| Flavor Used 			     | kubevirt VM           |
+| Distribution	                     | IDO       	     |
```
Comment (archie951) [inline applications/picocdn-node-vm/README.md:38]: Can you please change the distribution to Opensource.

Comment (piconets-engineering) [reply applications/picocdn-node-vm/README.md]: Done.

**Review by archie951 — APPROVED** [review]

**Review by abjyoti — APPROVED** [review]

**Comment by archie951** [comment]
Hi picoNETS team,
Can you please click on Update branch on your PR so that the Travis run sequence can get trigger again.

**Comment by piconets-engineering** [comment]
Hi All,

Just did an update on our fork, and as you said, it triggered a travisCI
auto build. That build appears to have passed, and the PR is awaiting final
merge.

Regards
R. K. Rajeev

**Pull request merged**